src/System.Text.Json/src/System/Text/Json/Document/JsonDocumentOptions.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json
+{
+    /// <summary>
+    /// Provides the ability for the user to define custom behavior when parsing JSON to create a <see cref="JsonDocument"/>.
+    /// </summary>
+    public struct JsonDocumentOptions
+    {
+        internal const int DefaultMaxDepth = 64;
+
+        private int _maxDepth;
+
+        /// <summary>
+        /// Defines how the <see cref="Utf8JsonReader"/> should handle comments when reading through the JSON.
+        /// By default <exception cref="JsonException"/> is thrown if a comment is encountered.
+        /// </summary>
+        public JsonCommentHandling CommentHandling { get; set; }
+
+        /// <summary>
+        /// Gets or sets the maximum depth allowed when reading JSON, with the default (i.e. 0) indicating a max depth of 64.
+        /// Reading past this depth will throw a <exception cref="JsonException"/>.
+        /// </summary>
+        public int MaxDepth
+        {
+            get => _maxDepth;
+            set
+            {
+                if (value < 0)
+                    throw ThrowHelper.GetArgumentException_MaxDepthMustBePositive();
```

Comment:
I've asked this on other PRs but I don't think I've seen an answer.  Why are we using this `throw ThrowHelper.Get*` approach rather than either just `throw new ArgumentException(...)` or `ThrowHelper.Throw*`?  I'm not understanding the purpose of the middle-ground.

Reply:
I had responded previously, but it probably got missed because it was within a comment thread that was "resolved".

Essentially, where the throwing is happening right at the top of a public method as part of argument validation, there is (arguably) _some_ value in keeping the `throw` keyword in the caller. 

https://github.com/dotnet/corefx/pull/38469#discussion_r294597102

I don't think we are consistent across the board here. I think, depending on who had written it/what they valued more at the time, they picked one or the other and the rationale for one or the other isn't super concrete (at least as far as I remember). I had used both myself (i.e. wasn't always consistent).

There are two patterns:
1) `throw HelperToGetException(...)`
   - Used to build up an exception with an exception message where building up the message is non-trivially expensive (or if there are multiple arguments and we need to avoid extra comparisons).
   - Not absolutely necessary, pattern 2 can be used just as well for most of these cases.
   - The benefit here is that the `throw` keyword is clearly visible in the caller (and so it's obvious that we aren't returning from this point).
2) `ThrowCallingHelperWhichThrows()`
   - Used when there isn't a single exception type that can be thrown, so a Helper that returns an exception wouldn't be feasible (however, you can also just return the base type `Exception` in pattern 1).
   - Moving the `throw` into the helper makes the caller code smaller.
   - Consistent with the existing pattern: 
https://github.com/dotnet/corefx/blob/fb81332b9643ecf8c0bb34d7938da97aa5e13e03/src/Common/src/CoreLib/System/ThrowHelper.cs#L86-L90
https://github.com/dotnet/corefx/blob/fb81332b9643ecf8c0bb34d7938da97aa5e13e03/src/System.Memory/src/System/ThrowHelper.cs#L27-L29

Ideally, we would just use pattern 2 anywhere it made sense to do so. In some cases, where the throwing is happening right at the top of a public method as part of argument validation, there is (arguably) _some_ value in keeping the `throw` keyword in the caller. That's probably why we have pattern 1 being used in this instance.

[sharplab.io for example](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzNNdzJA2AQIkhgFlyACl48GCAJTHT5/7wAZgwuCAwAPAx0Xn7+ZibUcUkMGAAWUBAA7gxcMNkAklwK2JK8ACYA8mIy2HwQXACiCCpidVwuACQARAK4uNgA5jAgxggAvl1eGonJY7FxxADsngwApAz6ALybUdP+czNm8zpM5Na29k6kbh7evodxQSFhkdHz/gnJcWkZ2QDiMAwhWKpUq1SgtV49SaLTaoSm73MBy+SxW6y2OzoeyR82OljONjsDkc2huGE8Pk+ySeoQiURiDw+iKSABV0llgSVylUam0YTBWlD2t5sUlkclUWF0Qxtrt5uKjg8ANqOQGpCBlfLiSQuVVpDVasSSKptXAsAByEEKpS47gGXgAuvMxFBeMUMDBTtZOaCeRC+c0BW0GACgUUuWDeUL+YL6mSKfMqUlUbkCuHfeDIdDA7H2t1ev0hiMjONJqKGAqGM7Xe7PVYmCgGGzfj7uZmA7ChfHvInmSl2f9Aa3I/7ozm4SL5TQxkA===)

Reply:
> In some cases, where the throwing is happening right at the top of a public method as part of argument validation, there is (arguably) some value in keeping the throw keyword in the caller

What value?  Just that the call site has "throw" instead of "Throw"?  I'm not sure why that's more beneficial when doing argument validation; if anything I'd think it'd be more beneficial when nestled inside the method, where throws are less common as compared to argument validation.

Regardless, my preference is for consistency here, and if we're going to use ThrowHelper, which is all about perf, being consistent on the form that eeks out the best perf, aka (2).

Reply:
> What value? Just that the call site has "throw" instead of "Throw"?

That and the control flow is clear (especially where the compiler would complain about things like out parameters not set).

> Regardless, my preference is for consistency here, and if we're going to use ThrowHelper, which is all about perf, being consistent on the form that eeks out the best perf, aka (2).

Agreed.